clients/include/testing_gebsr2csr.hpp

```diff
@@ -0,0 +1,626 @@
+                                     dbsr_val,
+                                     dbsr_row_ptr,
+                                     dbsr_col_ind,
+                                     row_block_dim,
+                                     col_block_dim,
+                                     csr_descr,
+                                     dcsr_val,
+                                     dcsr_row_ptr,
+                                     dcsr_col_ind);
+
+        if(mb < 0 || nb < 0 || row_block_dim < 0 || col_block_dim < 0)
+        {
+            verify_hipsparse_status_invalid_size(
+                status, "Error: mb < 0 || nb < 0 || row_block_dim < 0 || col_block_dim < 0");
+        }
+        else
+        {
+            verify_hipsparse_status_success(
+                status, "mb >= 0 && nb >= 0 && row_block_dim >= 0 && col_block_dim >= 0");
+        }
+
+        return HIPSPARSE_STATUS_SUCCESS;
+    }
+
+    // Read or construct CSR matrix
+    std::vector<int> bsr_row_ptr;
+    std::vector<int> bsr_col_ind;
+    std::vector<T>   bsr_val;
+    int              nnzb;
+    srand(12345ULL);
+    if(binfile != "")
+    {
+        if(read_bin_matrix(
+               binfile.c_str(), mb, nb, nnzb, bsr_row_ptr, bsr_col_ind, bsr_val, bsr_idx_base)
+           != 0)
+        {
+            fprintf(stderr, "Cannot open [read] %s\n", binfile.c_str());
+            return HIPSPARSE_STATUS_INTERNAL_ERROR;
+        }
+    }
+    else if(argus.laplacian)
+    {
+        mb = nb
+            = gen_2d_laplacian(argus.laplacian, bsr_row_ptr, bsr_col_ind, bsr_val, bsr_idx_base);
+        nnzb = bsr_row_ptr[m];
+    }
+    else
+    {
+        std::vector<int> coo_row_ind;
+
+        if(filename != "")
+        {
+            if(read_mtx_matrix(
+                   filename.c_str(), mb, nb, nnzb, coo_row_ind, bsr_col_ind, bsr_val, bsr_idx_base)
+               != 0)
+            {
+                fprintf(stderr, "Cannot open [read] %s\n", filename.c_str());
+                return HIPSPARSE_STATUS_INTERNAL_ERROR;
+            }
+        }
+        else
+        {
+            double scale = 0.02;
+            if(m > 1000 || n > 1000)
+            {
+                scale = 2.0 / std::max(mb, nb);
+            }
+            nnzb = mb * scale * nb;
+            gen_matrix_coo(mb, nb, nnzb, coo_row_ind, bsr_col_ind, bsr_val, bsr_idx_base);
+        }
+
+        // Convert COO to CSR
+        bsr_row_ptr.resize(mb + 1, 0);
+        for(int i = 0; i < nnzb; ++i)
+        {
+            ++bsr_row_ptr[coo_row_ind[i] + 1 - bsr_idx_base];
+        }
+
+        bsr_row_ptr[0] = bsr_idx_base;
+        for(int i = 0; i < mb; ++i)
+        {
+            bsr_row_ptr[i + 1] += bsr_row_ptr[i];
+        }
+    }
+
+    m       = mb * row_block_dim;
+    n       = nb * col_block_dim;
+    int nnz = nnzb * row_block_dim * col_block_dim;
+    // Now use the csr matrix as the symbolic for the gebsr matrix.
+    bsr_val.resize(nnz);
+    int idx = 0;
```

Comment:
Might be good to move this into a utility routine or something so that it can be reused in the other gebsr rountines?

Reply:
can be done when we will need to reuse. (?)

Reply:
next batch.